// @comunica/actor-query-operation-update-create/lib/ActorQueryOperationCreate.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ActorQueryOperationCreate = void 0;
const bus_query_operation_1 = require("@comunica/bus-query-operation");
/**
 * A [Query Operation](https://github.com/comunica/comunica/tree/master/packages/bus-query-operation) actor that
 * handles SPARQL create operations.
 */
class ActorQueryOperationCreate extends bus_query_operation_1.ActorQueryOperationTypedMediated {
    constructor(args) {
        super(args, 'create');
    }
    async testOperation(pattern, context) {
        bus_query_operation_1.ActorQueryOperation.throwOnReadOnly(context);
        return true;
    }
    async runOperation(pattern, context) {
        // Delegate to update-quads bus
        const { updateResult } = await this.mediatorUpdateQuads.mediate({
            createGraphs: {
                graphs: [pattern.source],
                requireNonExistence: !pattern.silent,
            },
            context,
        });
        return {
            type: 'update',
            updateResult,
        };
    }
}
exports.ActorQueryOperationCreate = ActorQueryOperationCreate;
