// componentsjs/lib/preprocess/constructorargumentsmapping/ConstructorArgumentsElementMappingHandlerKeyValue.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConstructorArgumentsElementMappingHandlerKeyValue = void 0;
const ErrorResourcesContext_1 = require("../../util/ErrorResourcesContext");
/**
 * Handler for static key-value entries.
 */
class ConstructorArgumentsElementMappingHandlerKeyValue {
    canHandle(configRoot, constructorArgs, configElement, mapper) {
        return Boolean((constructorArgs.property.value || constructorArgs.property.valueRawReference) &&
            !constructorArgs.property.collectEntries);
    }
    handle(configRoot, constructorArgs, configElement, mapper) {
        if (constructorArgs.property.key) {
            // Throw if our key is not a literal
            if (constructorArgs.property.key.type !== 'Literal') {
                throw new ErrorResourcesContext_1.ErrorResourcesContext(`Detected illegal IRI object key, which is only allowed with collectEntries`, {
                    objectKey: constructorArgs.property.key.term.value,
                    constructorArgs,
                    config: configRoot,
                });
            }
            // Key-value
            return this.handleKeyValue(configRoot, constructorArgs, configElement, mapper);
        }
        // Only value
        return this.handleValue(configRoot, constructorArgs, configElement, mapper);
    }
    handleKeyValue(configRoot, constructorArgs, configElement, mapper) {
        const ret = mapper.objectLoader.createCompactedResource({});
        ret.property.key = constructorArgs.property.key;
        for (const value of mapper.getParameterValue(configRoot, constructorArgs.property.value || constructorArgs.property.valueRawReference, configElement, Boolean(constructorArgs.property.valueRawReference))) {
            ret.properties.value.push(value);
        }
        return [ret];
    }
    handleValue(configRoot, constructorArgs, configElement, mapper) {
        return mapper.getParameterValue(configRoot, constructorArgs.property.value || constructorArgs.property.valueRawReference, configElement, Boolean(constructorArgs.property.valueRawReference));
    }
}
exports.ConstructorArgumentsElementMappingHandlerKeyValue = ConstructorArgumentsElementMappingHandlerKeyValue;
