// componentsjs/lib/rdf/Iris.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IRIS_OWL = exports.PREFIX_OWL = exports.IRIS_DOAP = exports.PREFIX_DOAP = exports.IRIS_XSD = exports.PREFIX_XSD = exports.IRIS_RDFS = exports.PREFIX_RDFS = exports.IRIS_RDF = exports.PREFIX_RDF = exports.PREFIX_OM = exports.IRIS_OO = exports.PREFIX_OO = void 0;
const definePrefix = (prefix) => (suffix) => `${prefix}${suffix}`;
exports.PREFIX_OO = definePrefix('https://linkedsoftwaredependencies.org/vocabularies/object-oriented#');
exports.IRIS_OO = {
    Module: exports.PREFIX_OO('Module'),
    Class: exports.PREFIX_OO('Class'),
    AbstractClass: exports.PREFIX_OO('AbstractClass'),
    ComponentInstance: exports.PREFIX_OO('ComponentInstance'),
    component: exports.PREFIX_OO('component'),
    componentPath: exports.PREFIX_OO('componentPath'),
    parameter: exports.PREFIX_OO('parameter'),
};
exports.PREFIX_OM = definePrefix('https://linkedsoftwaredependencies.org/vocabularies/object-mapping#');
exports.PREFIX_RDF = definePrefix('http://www.w3.org/1999/02/22-rdf-syntax-ns#');
exports.IRIS_RDF = {
    subject: exports.PREFIX_RDF('subject'),
    object: exports.PREFIX_RDF('object'),
    type: exports.PREFIX_RDF('type'),
    JSON: exports.PREFIX_RDF('JSON'),
};
exports.PREFIX_RDFS = definePrefix('http://www.w3.org/2000/01/rdf-schema#');
exports.IRIS_RDFS = {
    imports: exports.PREFIX_RDFS('seeAlso'),
};
exports.PREFIX_XSD = definePrefix('http://www.w3.org/2001/XMLSchema#');
exports.IRIS_XSD = {
    string: exports.PREFIX_XSD('string'),
    boolean: exports.PREFIX_XSD('boolean'),
    integer: exports.PREFIX_XSD('integer'),
    number: exports.PREFIX_XSD('number'),
    int: exports.PREFIX_XSD('int'),
    byte: exports.PREFIX_XSD('byte'),
    long: exports.PREFIX_XSD('long'),
    float: exports.PREFIX_XSD('float'),
    decimal: exports.PREFIX_XSD('decimal'),
    double: exports.PREFIX_XSD('double'),
};
exports.PREFIX_DOAP = definePrefix('http://usefulinc.com/ns/doap#');
exports.IRIS_DOAP = {
    name: exports.PREFIX_DOAP('name'),
};
exports.PREFIX_OWL = definePrefix('http://www.w3.org/2002/07/owl#');
exports.IRIS_OWL = {
    Restriction: exports.PREFIX_OWL('Restriction'),
};
