// @comunica/actor-query-operation-update-clear/lib/ActorQueryOperationClear.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ActorQueryOperationClear = void 0;
const bus_query_operation_1 = require("@comunica/bus-query-operation");
const rdf_data_factory_1 = require("rdf-data-factory");
const DF = new rdf_data_factory_1.DataFactory();
/**
 * A [Query Operation](https://github.com/comunica/comunica/tree/master/packages/bus-query-operation) actor
 * that handles SPARQL clear operations.
 */
class ActorQueryOperationClear extends bus_query_operation_1.ActorQueryOperationTypedMediated {
    constructor(args) {
        super(args, 'clear');
    }
    async testOperation(pattern, context) {
        bus_query_operation_1.ActorQueryOperation.throwOnReadOnly(context);
        return true;
    }
    async runOperation(pattern, context) {
        // Delegate to update-quads bus
        let graphs;
        if (pattern.source === 'DEFAULT') {
            graphs = DF.defaultGraph();
        }
        else if (typeof pattern.source === 'string') {
            graphs = pattern.source;
        }
        else {
            graphs = [pattern.source];
        }
        const { updateResult } = await this.mediatorUpdateQuads.mediate({
            deleteGraphs: {
                graphs,
                requireExistence: !pattern.silent,
                dropGraphs: false,
            },
            context,
        });
        return {
            type: 'update',
            updateResult,
        };
    }
}
exports.ActorQueryOperationClear = ActorQueryOperationClear;
