// @comunica/actor-query-operation-update-move-rewrite/lib/ActorQueryOperationMoveRewrite.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ActorQueryOperationMoveRewrite = void 0;
const bus_query_operation_1 = require("@comunica/bus-query-operation");
const sparqlalgebrajs_1 = require("sparqlalgebrajs");
/**
 * A [Query Operation](https://github.com/comunica/comunica/tree/master/packages/bus-query-operation) actor that
 * handles SPARQL move operations.
 */
class ActorQueryOperationMoveRewrite extends bus_query_operation_1.ActorQueryOperationTypedMediated {
    constructor(args) {
        super(args, 'move');
        this.factory = new sparqlalgebrajs_1.Factory();
    }
    async testOperation(pattern, context) {
        bus_query_operation_1.ActorQueryOperation.throwOnReadOnly(context);
        return true;
    }
    runOperation(pattern, context) {
        // No-op if source === destination
        if ((typeof pattern.destination === 'string' && typeof pattern.source === 'string' &&
            pattern.destination === pattern.source) ||
            (typeof pattern.destination !== 'string' && typeof pattern.source !== 'string' &&
                pattern.destination.equals(pattern.source))) {
            return Promise.resolve({
                type: 'update',
                updateResult: Promise.resolve(),
            });
        }
        // MOVE is equivalent to drop destination, add, and drop source
        const updates = [
            this.factory.createDrop(pattern.destination, true),
            this.factory.createAdd(pattern.source, pattern.destination, pattern.silent),
            this.factory.createDrop(pattern.source),
        ];
        const operation = this.factory.createCompositeUpdate(updates);
        return this.mediatorQueryOperation.mediate({ operation, context });
    }
}
exports.ActorQueryOperationMoveRewrite = ActorQueryOperationMoveRewrite;
